### PR TITLE
chore(client): [DPGA] implement --data-proxy

### PR DESCRIPTION
**Pull request by millsp**
This PR replaces the `PRISMA_CLIENT_ENGINE_TYPE=dataproxy` mechanism in favor of `--data-proxy` via the cli.

Next https://github.com/prisma/prisma/pull/13512

https://github.com/orgs/prisma/projects/70#card-82299232

### CODE REVIEW

**Comment by SevInf** [comment]
Warning: potentially stupid question ahead. 
Should we do something in case when `engineType` defined explicitly and `--data-proxy` flag is used? I am not sure what hypothetical user wants to achieve with this combination, but `engineType` would be ignored in this case, so maybe it worth a warning?

File: helpers/compile/build.ts [hunk 1]
```diff
@@ -183,7 +183,7 @@ const watch =
     }, 10)
 
     // triggers a full rebuild on added file
-    const onAdd = debounce(async () => {
+    const onAddOrUnlink = debounce(async () => {
```
Comment (millsp) [inline helpers/compile/build.ts:186]: Unrelated change to restart esbuild compiler on file deletion.

File: packages/cli/src/Generate.ts [hunk 2]
```diff
@@ -49,9 +47,10 @@ ${chalk.bold('Usage')}
 
 ${chalk.bold('Options')}
 
-  -h, --help   Display this help message
-    --schema   Custom path to your Prisma schema
-     --watch   Watch the Prisma schema and rerun after a change
+    -h, --help   Display this help message
+      --schema   Custom path to your Prisma schema
+  --data-proxy   Enable the Data Proxy in the Prisma Client
```
Comment (millsp) [inline packages/cli/src/Generate.ts:52]: New way to enable `--data-proxy`. There will be a followup later to also enable this for `prisma studio`.

File: packages/cli/src/Generate.ts [hunk 3]
```diff
@@ -137,6 +137,7 @@ ${chalk.bold('Examples')}
         printDownloadProgress: !watchMode,
         version: enginesVersion,
         cliVersion: pkg.version,
+        dataProxy: !!args['--data-proxy'],
```
Comment (millsp) [inline packages/cli/src/Generate.ts:140]: Pass the info down to the generators.

File: packages/client/fixtures/dataproxy/prisma/schema.prisma [hunk 4]
```diff
@@ -5,7 +5,6 @@ datasource db {
 
 generator client {
   provider   = "prisma-client-js"
-  engineType = "dataproxy"
```
Comment (millsp) [inline packages/client/fixtures/dataproxy/prisma/schema.prisma:8]: Not a valid entry anymore.

File: packages/client/src/__tests__/buildNFTAnnotations.test.ts [hunk 5]
```diff
@@ -11,7 +11,7 @@ function normalizePaths(snapshot: string): string {
 
 describe('library', () => {
   it('generates annotations for a schema and a single engine', () => {
-    const annotations = buildNFTAnnotations(ClientEngineType.Library, ['debian-openssl-1.1.x'], 'out')
+    const annotations = buildNFTAnnotations(false, ClientEngineType.Library, ['debian-openssl-1.1.x'], 'out')
```
Comment (millsp) [inline packages/client/src/__tests__/buildNFTAnnotations.test.ts:14]: `dataProxy = false`

File: packages/client/src/__tests__/buildNFTAnnotations.test.ts [hunk 6]
```diff
@@ -84,16 +86,15 @@ describe('binary', () => {
 describe('dataproxy', () => {
   it('generates no annotations', () => {
     const annotations = buildNFTAnnotations(
-      ClientEngineType.DataProxy,
+      true,
```
Comment (millsp) [inline packages/client/src/__tests__/buildNFTAnnotations.test.ts:89]: `dataProxy = true`

File: packages/client/src/__tests__/integration/happy/custom-engine-binary/test.ts [hunk 7]
```diff
@@ -5,11 +5,6 @@ import path from 'path'
 
 import { generateTestClient } from '../../../../utils/getTestClient'
 
-if (getClientEngineType() === ClientEngineType.DataProxy) {
```
Comment (millsp) [inline packages/client/src/__tests__/integration/happy/custom-engine-binary/test.ts:8]: Not needed.

File: packages/client/src/generation/TSClient/TSClient.ts [hunk 8]
```diff
@@ -86,15 +88,16 @@ export class TSClient implements Generatable {
       engineVersion: this.options.engineVersion,
       datasourceNames: datasources.map((d) => d.name),
       activeProvider: this.options.activeProvider,
+      dataProxy: this.options.dataProxy,
```
Comment (millsp) [inline packages/client/src/generation/TSClient/TSClient.ts:91]: `dataProxy` is now stored in the generated config instead of being in the generator config block (via schema).

File: packages/client/src/generation/TSClient/TSClient.ts [hunk 9]
```diff
@@ -62,6 +63,7 @@ export class TSClient implements Generatable {
       runtimeDir,
       runtimeName,
       datasources,
+      dataProxy,
```
Comment (millsp) [inline packages/client/src/generation/TSClient/TSClient.ts:66]: Passed by the generator.

File: packages/client/src/generation/TSClient/TSClient.ts [hunk 10]
```diff
@@ -112,22 +115,22 @@ ${new Enum(
...
 const config = ${JSON.stringify(config, null, 2)}
 config.document = dmmf
 config.dirname = dirname
-${buildInlineDatasource(engineType, datasources)}
-${await buildInlineSchema(engineType, schemaPath)}
-${buildInlineEnv(engineType, datasources, envPaths)}
-${buildWarnEnvConflicts(engineType, runtimeDir, runtimeName)}
+${buildInlineDatasource(dataProxy, datasources)}
+${await buildInlineSchema(dataProxy, schemaPath)}
+${buildInlineEnv(dataProxy, datasources, envPaths)}
```
Comment (millsp) [inline packages/client/src/generation/TSClient/TSClient.ts:128]: This will be removed in the followup PR.

File: packages/client/src/generation/TSClient/TSClient.ts [hunk 11]
```diff
@@ -86,15 +88,16 @@ export class TSClient implements Generatable {
...
     }
 
     // get relative output dir for it to be preserved even after bundling, or
     // being moved around as long as we keep the same project dir structure.
     const relativeOutdir = path.relative(process.cwd(), outputDir)
 
     const code = `${commonCodeJS({ ...this.options, browser: false })}
-${buildRequirePath(engineType)}
-${buildDirname(engineType, relativeOutdir, runtimeDir)}
+${buildRequirePath(dataProxy)}
```
Comment (millsp) [inline packages/client/src/generation/TSClient/TSClient.ts:99]: We don't use the engine type in most cases anymore as that was used to do the data proxy distinction at generation time.

File: packages/client/src/generation/generateClient.ts [hunk 12]
```diff
@@ -57,7 +58,6 @@ export interface BuildClientResult {
 
 // eslint-disable-next-line @typescript-eslint/require-await
 export async function buildClient({
-  datamodel,
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:60]: Not used.

File: packages/client/src/generation/generateClient.ts [hunk 13]
```diff
@@ -48,6 +48,7 @@ export interface GenerateClientOptions {
   engineVersion: string
   clientVersion: string
   activeProvider: string
+  dataProxy: boolean
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:51]: Passed to `generateClient` via the generator.

File: packages/client/src/generation/generateClient.ts [hunk 14]
```diff
@@ -238,7 +242,8 @@ export async function generateClient({
       }\` in the \`binaryPaths\` object.`,
     )
   }
-  if (transpile) {
+
+  if (transpile === true && dataProxy === false) {
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:246]: We don't copy the query engine files if the data proxy is in use.

File: packages/client/src/generation/generator.ts [hunk 15]
```diff
@@ -51,6 +51,7 @@ if (process.argv[1] === __filename) {
         clientVersion,
         transpile: true,
         activeProvider: options.datasources[0]?.activeProvider,
+        dataProxy: options.dataProxy,
```
Comment (millsp) [inline packages/client/src/generation/generator.ts:54]: Passed by the cli.

File: packages/client/src/runtime/getPrismaClient.ts [hunk 16]
```diff
@@ -449,7 +458,9 @@ export function getPrismaClient(config: GetPrismaClientConfig) {
       return 'PrismaClient'
     }
     private getEngine(): Engine {
-      if (this._clientEngineType === ClientEngineType.Library) {
+      if (this._dataProxy === true) {
```
Comment (millsp) [inline packages/client/src/runtime/getPrismaClient.ts:461]: `dataProxy = true` is an override of the other engines, as explicitly asked via `--data-proxy`

File: packages/client/tests/functional/typescript/tests.ts [hunk 17]
```diff
@@ -36,7 +36,7 @@ function getAllTestSuiteTypeChecks(fileNames: string[]) {
 
 describe('typescript', () => {
   const suitePaths = glob.sync('./**/.generated/**/*.ts', {
-    ignore: ['./**/.generated/**/*.d.ts', './**/.generated/**/_schema.ts'],
+    ignore: ['./**/.generated/**/*.d.ts', './**/.generated/**/_*.ts'],
```
Comment (millsp) [inline packages/client/tests/functional/typescript/tests.ts:39]: Unrelated change.

File: packages/debug/helpers/build.ts [hunk 18]
```diff
@@ -1,3 +1,3 @@
 import { build } from '../../../helpers/compile/build'
 
-void build([{ name: 'default' }])
+void build([{ name: 'debug' }])
```
Comment (millsp) [inline packages/debug/helpers/build.ts:3]: Unrelated change.

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts [hunk 19]
```diff
@@ -237,7 +234,7 @@ export class DataProxyEngine extends Engine {
     const { protocol, host, searchParams } = url
 
     if (protocol !== 'prisma:') {
-      throw new InvalidDatasourceError('Datasource URL should use prisma:// protocol. If you are not using the Data Proxy, remove the `dataProxy` from the `previewFeatures` in your schema and ensure that `PRISMA_CLIENT_ENGINE_TYPE` environment variable is not set to `dataproxy`.', {
+      throw new InvalidDatasourceError('Datasource URL should use prisma:// protocol when --data-proxy is used', {
```
Comment (millsp) [inline packages/engine-core/src/data-proxy/DataProxyEngine.ts:237]: TODO: use "must" instead of "should"

File: packages/engine-core/src/data-proxy/utils/request.ts [hunk 20]
```diff
@@ -1,20 +1,24 @@
...
 
-// our implementation handles less
-export type RequestOptions = O.Patch<
-  {
-    headers?: { [k: string]: string }
-    body?: string
-  },
-  RequestInit
->
+// our implementation handles less than the real fetch
```
Comment (millsp) [inline packages/engine-core/src/data-proxy/utils/request.ts:6]: Because of the removal of ambient types from `webworkers`.

File: packages/migrate/src/Migrate.ts [hunk 21]
```diff
@@ -155,6 +155,7 @@ export class Migrate {
       printDownloadProgress: true,
       version: enginesVersion,
       cliVersion: packageJson.version,
+      dataProxy: false,
```
Comment (millsp) [inline packages/migrate/src/Migrate.ts:158]: We don't allow `--data-proxy` on `prisma migrate` as they are not currently compatible.

File: packages/sdk/src/__tests__/getGenerators/getGenerators.test.ts [hunk 22]
```diff
@@ -761,17 +773,13 @@ describe('getGenerators', () => {
...
       expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
         "
-        The dataProxy and interactiveTransactions Preview Features can not be enabled at the same time.
-        Remove interactiveTransactions from previewFeatures, for example:
-
-        generator client {
-            provider = \\"prisma-client-js\\"
-            previewFeatures = [\\"dataProxy\\"]
-        }
+        interactiveTransactions preview feature is not yet available with --data-proxy.
```
Comment (millsp) [inline packages/sdk/src/__tests__/getGenerators/getGenerators.test.ts:781]: TODO: re-enable this test.

File: packages/sdk/src/client/getClientEngineType.ts [hunk 23]
```diff
@@ -3,7 +3,6 @@ import type { GeneratorConfig } from '@prisma/generator-helper'
 export enum ClientEngineType {
   Library = 'library',
   Binary = 'binary',
-  DataProxy = 'dataproxy',
```
Comment (millsp) [inline packages/sdk/src/client/getClientEngineType.ts:6]: Not allowed here anymore.

File: packages/sdk/src/get-generators/utils/check-feature-flags/checkFeatureFlags.ts [hunk 24]
```diff
@@ -1,33 +1,26 @@
...
+/**
+ * Check feature flags and preview features
+ * @param config
+ * @param options
+ */
+export function checkFeatureFlags(config: ConfigMetaFormat, options: GetGeneratorOptions) {
+  checkForbiddenItxWithDataProxyFlag(config, options)
 }
 
-function checkProxyFeatureFlag(config: ConfigMetaFormat) {
```
Comment (millsp) [inline packages/sdk/src/get-generators/utils/check-feature-flags/checkFeatureFlags.ts:10]: Not needed anymore.

File: packages/sdk/src/get-generators/utils/check-feature-flags/checkFeatureFlags.ts [hunk 25]
```diff
@@ -1,33 +1,26 @@
...
+function checkForbiddenItxWithDataProxyFlag(config: ConfigMetaFormat, options: GetGeneratorOptions) {
```
Comment (millsp) [inline packages/sdk/src/get-generators/utils/check-feature-flags/checkFeatureFlags.ts:14]: Make sure that itx cannot be used with --data-proxy.

**Comment by millsp** [comment]
Replaced by https://github.com/prisma/prisma/pull/13561
